feat(dhall-docs): multi-page docs generation

german1608 opened the pull request:
This attempts to generate a really simple and non-flavoured HTML. There is no CSS or JS, just HTML. This doesn't include markdown parsing.

I researched about libraries to parse markdown and I found the two following:

* [`mmark`]
* [`cmark`](https://hackage.haskell.org/package/cmark)

Any of them let us extend its markdown parser, so new elements will need to be parsed after calling their functions. I haven't decided yet what library to use, but soon I'll do.

Also, I researched about libraries to generate HTML and I found:

* [`lucid`]
* [`blaze-html`](https://hackage.haskell.org/package/blaze-html)

I ended up using [`lucid`] because:

* (objectively) [`mmark`] outputs the `Html` data type provided by it.
* (subjectively) I feel more confortable using it vs `blaze-html`

[`mmark`]: https://hackage.haskell.org/package/mmark
[`lucid`]: https://hackage.haskell.org/package/lucid

german1608 commented:
Currently a draft because the generate documentation doesn't have an index.html so its really awkard to navigate on it

german1608 commented:
There are no tests because I don't know how to test this. Maybe a golden-test setup like we did on #1799?

german1608 commented:
Well, I think that this is ready to be reviewed.

Summary of what's exactly included:

* Multi-page HTML generation without markdown parsing.
* New parameter for command line `--output`, that defaults to `./docs/`
* A simple index.css file and some fonts ([Roboto mono](https://fonts.google.com/specimen/Roboto+Mono?query=roboto+mono&sidebar.open&selection.family=Lato|Roboto+Mono#standard-styles) and [Lato](https://fonts.google.com/specimen/Lato?query=lato&sidebar.open&selection.family=Lato|Roboto+Mono#standard-styles)) that just changes the font-family of all html elements.

... and a list of known bugs that probably with your codereview I can fix:

* The `resolveRelativePath` is used to properly link the css in all the html pages, but its buggy:
  + if `outDir` doesn't ends in `/` or isn't properly relative, then it loops endlesly. It would be great if you know of a function that takes two directories (one inside the other) and returns the number of `..` needed to go there that takes into account path separators and other OS-dependant facts

sjakobi commented:
Amazing! I haven't taken a proper look at the code, but I have some comments on the dependencies and the fonts:

One worrying thing about `lucid` is that its maintainer recently turned off its issue tracker at https://github.com/chrisdone/lucid. I'm not sure what that means for the future of the package. It might be worth starting a thread on the libraries mailing list or r/haskell and inquire whether there's anyone committed to maintaining it in the future.

---

Regarding Markdown libraries:

This has implications for the flavour of Markdown we support and possible extensions, for the difficulty of specifying the documentation syntax, and the possibility of parsing the docs with other tools.

If we want to achieve good compatibility, we should probably stick as closely as possible to CommonMark.

The nice thing about `mmark` is its strictness, which should make it easier for users to avoid syntax errors.

I was also thinking about https://github.com/jgm/commonmark-hs which has been maturing for quite a while but unfortunately doesn't have proper release yet. Maybe inquire about that? The author might also have some advice on our choice of libraries – he's quite the expert on the matter.

Lots of details to consider here… @german1608 Could you possibly make a new issue and try to lay out the pros and cons?

---

Regarding the fonts: Do we have to consider any licensing issues when we include these fonts in the package?

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+            isNotDir <- not <$> Turtle.testdir path_
+            Control.Monad.guard isNotDir
+
+            let pathStr = Text.unpack $ Turtle.format fp path_
```
I'm not very familiar with `turtle`, so I'd prefer if e.g. `fp` too was imported qualified.

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+    -> IO [(FilePath, Header)]
+getAllDhallFiles baseDir = do
+
+    let shell = do
```
Can you add a type annotation?

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+
+    Turtle.fold shell Foldl.list
+
+{-| Calculate the relative path needed to access files on the first argument
+    relative from the second argument.
+
+    The second argument needs to be a child of the first, otherwise it will
+    loop forever
+-}
+resolveRelativePath :: FilePath -> FilePath -> FilePath
```
I find it hard to tell what this does from the description. Could you add examples, ideally as doctests?

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+    if outDir == currentDir then ""
+    else ".." </> resolveRelativePath (outDir </> "") (Turtle.parent currentDir)
+
+saveHtml :: FilePath -> FilePath -> (FilePath, Header) -> IO FilePath
+saveHtml inputPath outputPath t@(filePath, _) = do
```
The amount of `FilePath`s involved here is a bit confusing. If we do a lot of filepath munging, I'd suggest using [`path`](http://hackage.haskell.org/package/path) which generally makes this a bit clearer by adding a distinction between relative and absolute paths, and paths to directories and files.

sjakobi replied:
There's also a companion package for IO operations: http://hackage.haskell.org/package/path-io

german1608 replied:
Thanks for the suggestion. I'll take a look to it

german1608 commented:
@sjakobi: 
> Amazing! I haven't taken a proper look at the code, but I have some comments on the dependencies and the fonts:
> 
> One worrying thing about `lucid` is that its maintainer recently turned off its issue tracker at https://github.com/chrisdone/lucid. I'm not sure what that means for the future of the package. It might be worth starting a thread on the libraries mailing list or r/haskell and inquire whether there's anyone committed to maintaining it in the future.

Yes, I noticed it but didn't thought of that regard. Do you know where is the mailing list of that library? I can't go with reddit sadly, my only internet provider has it banned for some reason.

> 
> Regarding Markdown libraries:
> 
> This has implications for the flavour of Markdown we support and possible extensions, for the difficulty of specifying the documentation syntax, and the possibility of parsing the docs with other tools.
> 
> If we want to achieve good compatibility, we should probably stick as closely as possible to CommonMark.
> 
> The nice thing about `mmark` is its strictness, which should make it easier for users to avoid syntax errors.
> 
> I was also thinking about https://github.com/jgm/commonmark-hs which has been maturing for quite a while but unfortunately doesn't have proper release yet. Maybe inquire about that? The author might also have some advice on our choice of libraries – he's quite the expert on the matter.
> 
> Lots of details to consider here… @german1608 Could you possibly make a new issue and try to lay out the pros and cons?

Sure, I'll file an issue.

> 
> Regarding the fonts: Do we have to consider any licensing issues when we include these fonts in the package?

Roboto Mono uses [Apache 2.0](https://cd.mahisoft.com/job/Kamino/job/kamino-ms-file-ng/job/master/) 
Lato uses [Open Font](https://fonts.google.com/specimen/Lato?query=lato#license)
I'm not very fluent on licenses, are they compatible with our project one?

Gabriel439 commented on dhall-docs/README.md:
````diff
@@ -9,3 +9,18 @@ For installation or development instructions, see:
+The easiest usage is the following:
+
+```bash
+dhall-docs --input <YOUR-PACKAGE-FOLDER>
````
I like to use Bash variables as placeholders for user-supplied command-line values in documentation:

```suggestion
dhall-docs --input "${YOUR-PACKAGE-FOLDER}"
```

Gabriel439 commented on dhall-docs/README.md:
````diff
@@ -9,3 +9,18 @@ For installation or development instructions, see:
+you can change that using the `--output` flag
+
+```bash
+dhall-docs --input . --output other-dir
````
```suggestion
dhall-docs --input . --output "${OTHER-DIR}"
```

Gabriel439 commented on dhall-docs/dhall-docs.cabal:
```diff
@@ -22,26 +22,39 @@ Description:
     CHANGELOG.md
+Data-Files:
+    src/Dhall/data/*.css,
+    src/Dhall/data/*.ttf
```
Perhaps there could be an option to embed these data files in the executable (for ease of distributing standalone executables, such as on our Releases page).  `pandoc` and `criterion` both have an an `-fembed-data-files` configuration flag to do something similar, so you can study either one of them to see how it is done

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+        ( Options.Applicative.long "input"
+       <> Options.Applicative.metavar "INPUT"
+       <> Options.Applicative.help "Directory of your dhall package. It should \
+                                    \contain only text files" )
```
Is the text-only restriction necessary?  We could ignore files that don't parse as valid Dhall code

german1608 replied:
Well I left a TODO comment here:
https://github.com/dhall-lang/dhall-haskell/blob/470f9e359ec7443db7f1ff2955c7109d0655706d/dhall-docs/src/Dhall/Docs.hs#L92-L100

do you know of an util that checks if the file is binary? That gave me an error

Gabriel439 replied:
@german1608: You can read the file as a `ByteString` using `Data.ByteString.readFile` and then use `Data.Text.decodeUtf8'`.  Then if decoding fails you know the file is binary

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+{-| Calculate the relative path needed to access files on the first argument
+    relative from the second argument.
+
+    The second argument needs to be a child of the first, otherwise it will
```
I think a more reliable way to do this is to make the two paths absolute, delete the longest shared prefix between the two paths, and use the remainder to compute the relative path between the two

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +68,116 @@ parseOptions :: Parser Options
+            if directory `Map.notMember` accMap then Map.insert directory [file] accMap
+            else Map.adjust ((:) file) directory accMap
+
+    let filesGroupedByDir = foldl groupByDir Map.empty htmlFiles
```
A more high-level way to do this is:

```haskell
let toMap file = Map.singleton file [ Turtle.directory file ]

let filesGroupedByDir = Map.unionsWith (<>) (map toMap htmlFiles)
```

german1608 commented:
Lucky-me I managed to get some internet and push my latest work.

@Gabriel439 I added a flag to the dhall-docs.cabal: `ghci-data-files`.

I checked out how `criterion` embed its files for the project and they use [`file-embed`] for that matter. I configured it and worked properly! But when I wanted to run `stack ghci dhall-docs` to debug some new code it didn't resolved the path properly.

So I added that flag that, when supplied, will make an executable without the embedded data-files. My only purpose of that flag was to be able to execute `stack ghci dhall-docs`, so that give me the name `ghci-data-files`. `embed-data-files` worked as well but that is actutally the expected behaviour and I didn't liked the `no-embed-data-files` - too long.

I think that what is missing here is to try to use a more type-safe path management library, like @sjakobi suggested with [`path`]. This tool will make heavy use of filepaths and I think it will gain profit from using it.

[`file-embed`]: https://hackage.haskell.org/package/file-embed
[`path`]: https://hackage.haskell.org/package/path

german1608 commented:
The size of the executable with the embedded data-files didn't change a lot: 14MB embedded, 13MB non-embedded

sjakobi commented:
> > One worrying thing about `lucid` is that its maintainer recently turned off its issue tracker at https://github.com/chrisdone/lucid. I'm not sure what that means for the future of the package. It might be worth starting a thread on the libraries mailing list or r/haskell and inquire whether there's anyone committed to maintaining it in the future.
> 
> Yes, I noticed it but didn't thought of that regard. Do you know where is the mailing list of that library? I can't go with reddit sadly, my only internet provider has it banned for some reason.

This is the libraries mailing list: https://mail.haskell.org/cgi-bin/mailman/listinfo/libraries

I've already sent an [email](https://mail.haskell.org/pipermail/libraries/2020-June/030616.html) and CC'd you. I've also seen that phadej, who is a very prolific maintainer, has recently committed to the repo, which seems like a good sign.

german1608 commented:
> > > One worrying thing about `lucid` is that its maintainer recently turned off its issue tracker at https://github.com/chrisdone/lucid. I'm not sure what that means for the future of the package. It might be worth starting a thread on the libraries mailing list or r/haskell and inquire whether there's anyone committed to maintaining it in the future.
> > 
> > 
> > Yes, I noticed it but didn't thought of that regard. Do you know where is the mailing list of that library? I can't go with reddit sadly, my only internet provider has it banned for some reason.
> 
> This is the libraries mailing list: https://mail.haskell.org/cgi-bin/mailman/listinfo/libraries
> 
> I've already sent an [email](https://mail.haskell.org/pipermail/libraries/2020-June/030616.html) and CC'd you. I've also seen that phadej, who is a very prolific maintainer, has recently committed to the repo, which seems like a good sign.

Thanks @sjakobi , I'll make sure to bookmark that info

Gabriel439 commented on dhall-docs/README.md:
````diff
@@ -9,3 +9,18 @@ For installation or development instructions, see:
+dhall-docs --input ${PACKAGE-FOLDER}
+```
+
+By default it will save the documentation of your package in `$PWD/docs`, but
````
```suggestion
By default it will save the documentation of your package in `./docs`, but
```

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +69,114 @@ parseOptions :: Parser Options
+    successfully parses as a valid dhall file.
+
+    The reason it doesn't guide the search by its extension is because of the
+    dhall @Prelude@ (<https://github.com/dhall-lang/dhall-lang/tree/master/Prelude>)
```
Haddocks support display text for links, which you can use like this:

```suggestion
    dhall <https://prelude.dhall-lang.org Prelude>
```

I also shortened the Prelude link

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +69,114 @@ parseOptions :: Parser Options
+    createIndexes outDir generatedHtmlFiles
+
+    dataDir <- getDataDir
+    Control.Monad.forM_ dataDir $ \(filename, contents) ->
+        let finalPath = Turtle.encodeString $
+                outDir </> Turtle.decodeString filename
+        in
+        Data.ByteString.writeFile finalPath contents
```
Minor suggestion:

```suggestion
    Control.Monad.forM_ dataDir $ \(filename, contents) -> do
        let finalPath = Turtle.encodeString $
                outDir </> Turtle.decodeString filename

        Data.ByteString.writeFile finalPath contents
```

Gabriel439 commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +69,114 @@ parseOptions :: Parser Options
+                outDir </> Turtle.decodeString filename
+        in
+        Data.ByteString.writeFile finalPath contents
+    -- Turtle.cptree dataDir outDir
```
Did you mean to keep this comment?

Profpatsch commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +69,112 @@ parseOptions :: Parser Options
+    as finding all files that ends in @.dhall@, but finds all files that
+    successfully parses as a valid dhall file.
+
+    The reason it doesn't guide the search by its extension is because of the
+    dhall <https://prelude.dhall-lang.org Prelude>
+    That package doesn't ends any of their files in @.dhall@.
```
This is an interesting discussion to have: is this the point where we should switch to using `.dhall` extensions?

german1608 replied:
I thought of that but wouldn't that be a breaking change? Also I don't see any restriction on language specs that a valid dhall file must have a `.dhall` extension, and imposing it only for simplifyinig this tool can be a little overkill

Gabriel439 replied:
This is not necessarily a breaking change.  The Prelude could keep around the old extension-free files but make them re-export the ones with a `.dhall` extension

Profpatsch commented on dhall-docs/src/Dhall/data/index.css:
```diff
@@ -0,0 +1,110 @@
+}
+
+
+/* cyrillic-ext */
+
+@font-face {
+    font-family: 'Roboto Mono';
+    font-style: normal;
+    font-weight: 400;
+    font-display: swap;
+    src: local('Roboto Mono'), local('RobotoMono-Regular'), url(https://fonts.gstatic.com/s/robotomono/v7/L0x5DF4xlVMF-BfR8bXMIjhGq3-OXg.woff2) format('woff2');
+    unicode-range: U+0460-052F, U+1C80-1C88, U+20B4, U+2DE0-2DFF, U+A640-A69F, U+FE2E-FE2F;
```
I wonder what’s going on with all these `@font-face` sections? Is this usual boilerplate?

german1608 replied:
Usual boilerplate, `@font-face` is used to declare font-families in css

german1608 replied:
If you're curious, this was generate by google fonts when you select a set of fonts and download them. You get:

* The font files
* This css file with that boilerplate

german1608 commented:
@sjakobi I changed `FilePath` references to use `Path` from the `path` package. The places where I didn't do so where:
* On the `Options` ADT, since its better too use the `Path` ADT without parameterizing e.g. `Path Abs Dir` is better than `Path a b`, and the tool allow to use relative and absolute paths
* Places where I need to use `..` (`resolveRelativePath`, css imports refs in `Html.hs`, `Embedded.hs`)
* On `Embedded.hs` since the `embedDir` uses `FilePath` directly.

I'm working to see if I can change `Embedded.hs` to return `Path Rel File` instead.

german1608 commented:
Well now `getDataDir` from `Embedded.hs` returns `IO [(Path Rel File, ByteString)]`.

The following packages were removed from the dependencies:

- `turtle`
- `directory`
- `foldl`

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +72,157 @@ parseOptions :: Parser Options
+    -> (Path Abs File, Header)  -- ^ (Input file, Parsed header)
+    -> IO (Path Abs File)       -- ^ Output path file
+saveHtml inputAbsDir outputAbsDir t@(absFile, _) = do
+    htmlOutputFile <- (outputAbsDir </>)
+            <$> (Path.stripProperPrefix inputAbsDir absFile
+                >>= addHtmlExt)
```
I think a `do`-expression would probably be more readable here.

sjakobi commented on dhall-docs/src/Dhall/Docs.hs:
```diff
@@ -51,20 +72,157 @@ parseOptions :: Parser Options
+
+    Examples:
+
+    >>> resolveRelativePath $(mkAbsDir "/a/b/c/") $(mkAbsDir "/a/b/c/d/e/")
```
Personally, I think the QuasiQuoters look a bit nicer. It doesn't really matter though. ;)

```suggestion
    >>> resolveRelativePath [absdir|/a/b/c/] [absdir|/a/b/c/d/e/]
```

sjakobi commented:
Regarding the examples in the haddocks, I think it would be nice to get them checked automatically by `doctest`.

german1608 commented:
@sjakobi : I'll add your suggestions in a next PR, which will be either adding the markdown parsing or tests for the tool. Meanwhile, I'll merge this
